packages/mongo/mongo_driver.js

```diff
@@ -345,14 +341,12 @@ MongoConnection.prototype._insert = function (collection_name, document,
   }
 
   var write = self._maybeBeginWrite();
-  var refresh = function () {
-    Meteor.refresh({collection: collection_name, id: document._id });
-  };
-  callback = bindEnvironmentForWrite(writeCallback(write, refresh, callback));
+  callback = bindEnvironmentForWrite(writeCallback(write, callback));
   try {
     var collection = self.rawCollection(collection_name);
     collection.insert(replaceTypes(document, replaceMeteorAtomWithMongo),
                       {safe: true}, callback);
+    Meteor.refresh({collection: collection_name, id: document._id });
```

Comment:
I was operating on the assumption of 'read your own writes' consistency, but since different connection is used for oplog tailing, this might not be the case indeed. Reverted back to 'refresh on database response' behavior, and moved the fence callback from 'on armed' to 'on ready to fire' to accomodate for the fact that we need to wait for write replies.